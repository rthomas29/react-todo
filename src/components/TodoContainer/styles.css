.container {
  display: flex;
  flex-flow: column nowrap;
  justify-content: center;
  align-items: center;
  min-height: 250px;
  min-width: 250px;
  margin: auto;
  border: 1px solid #eeeeee;
  border-radius: 5px;
  padding: 10px;
  background-color: #fff;
  color: #6c7a89;
  box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2), 0 6px 20px 0 rgba(0, 0, 0, 0.19);
}