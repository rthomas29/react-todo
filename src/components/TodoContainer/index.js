import React, { useReducer, useEffect } from 'react'
import TodoForm from '../TodoForm'
import TodoList from '../TodoList'
import Title from '../Title'
import './styles.css'

const TodoContainer = () => {
  // fetch todoList from localStorage if it exists
  const initialState = { todoList: JSON.parse(localStorage.getItem('todoList')) || [] }

  // reducer to handle our state changes
  // created reducer to mirror setState behavior of class components
  const reducer = (state, newState) => ({ ...state, ...newState })
  const [state, setState] = useReducer(reducer, initialState)

  // push new task to todoList
  const handleTaskAdd = (task) => {
    if (task.length > 0) {
      const currentTodoList = [...state.todoList]
      const newTodo = { task, isCompleted: false }
      currentTodoList.push(newTodo)
      setState({ todoList: currentTodoList })
    }
  }

  // update todoList to only contain incomplete tasks
  const deleteCompletedTasks = () => {
    const currentTodoList = [...state.todoList]
    const remainingTasks = []
    for (let todo of currentTodoList) {
      // if the todo is not marked as complete, push to remaining tasks array
      if (!todo.isComplete) {
        remainingTasks.push(todo)
      }
    }
    // update todoList with remaingTasks array
    setState({ todoList: remainingTasks })
  }

  // toggle isComplete boolean for specified task
  const toggleTaskCompletion = (taskIndex) => {
    const currentTodoList = [...state.todoList]
    currentTodoList[taskIndex].isComplete = !currentTodoList[taskIndex].isComplete
    setState({ todoList: currentTodoList })
  }

  const hydrateTodoListFromLocalStorage = () => {
    if (localStorage.hasOwnProperty('todoList')) {
      const todoList = localStorage.getItem('todoList')
      try {
        setState({ todoList: JSON.parse(todoList) })
      } catch (error) {
        setState({ todoList: [] })
      }
    }
  }

  useEffect(() => {
    // whenever our state is updated, sync localStorage with app state
    localStorage.setItem('todoList', JSON.stringify(state.todoList))
  }, [state])

  const { todoList } = state
  const numTasks = todoList.length
  return (
    <div className='container'>
      <Title numTasks={numTasks} />
      <TodoForm handleTaskAdd={handleTaskAdd} />
      <TodoList
        todos={todoList}
        toggleTaskCompletion={toggleTaskCompletion}
        numTasks={numTasks}
        deleteCompletedTasks={deleteCompletedTasks}
      />
    </div>
  )
}

export default TodoContainer