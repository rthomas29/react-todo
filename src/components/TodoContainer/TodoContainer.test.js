import React from 'react'
import { render } from 'react-testing-library'
import TodoContainer from './index'

describe('<TodoContainer />', () => {
  it('should match snapshot', () => {
    const { container } = render(<TodoContainer />)
    expect(container.firstChild).toBeInTheDocument()
    expect(container.firstChild).toMatchSnapshot()
  })
})