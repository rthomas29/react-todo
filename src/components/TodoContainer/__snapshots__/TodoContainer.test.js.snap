// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoContainer /> should match snapshot 1`] = `
<section
  class="container"
>
  <h2>
    No tasks remaining ðŸ¤—
  </h2>
  <form
    class="form-inline"
  >
    <div
      class="form-group"
    >
      <label
        for="task-input"
      >
        Task:
      </label>
      <input
        aria-label="checkbox to toggle task completion"
        class="form-control"
        id="task-input"
        placeholder="Enter new task here"
        type="text"
        value=""
      />
      <button
        class="btn btn-success"
        type="submit"
      >
        Add
      </button>
    </div>
  </form>
  <main
    class="list-container"
  />
</section>
`;
